x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/CompositeAggCursor.java

```diff
@@ -162,71 +163,64 @@ static void handle(
+            updateCompositeAfterKey(response, source);
+            retry.run();
+            return;
+        }
 
-                if (afterKey != null) {
-                    updateSourceAfterKey(afterKey, source);
-                }
+        CompositeAggRowSet rowSet = makeRowSet.get();
 
-                Cursor next = rowSet.remainingData() == 0 ? Cursor.EMPTY : makeCursor.apply(source, rowSet);
-                listener.onResponse(new Page(rowSet, next));
-            } catch (Exception ex) {
-                listener.onFailure(ex);
-            }
-        }
-        // no results
-        else {
-            listener.onResponse(Page.last(Rows.empty(schema)));
+        Map<String, Object> afterKey = rowSet.afterKey();
+
+        if (afterKey != null) {
+            updateSourceAfterKey(afterKey, source);
         }
+
+        Cursor next = rowSet.remainingData() == 0 ? Cursor.EMPTY : makeCursor.apply(source, rowSet);
+        listener.onResponse(new Page(rowSet, next));
     }
 
     private static boolean shouldRetryDueToEmptyPage(SearchResponse response) {
         CompositeAggregation composite = getComposite(response);
         // if there are no buckets but a next page, go fetch it instead of sending an empty response to the client
-        return composite != null
-            && composite.getBuckets().isEmpty()
-            && composite.afterKey() != null
-            && composite.afterKey().isEmpty() == false;
+        return composite.getBuckets().isEmpty() && composite.afterKey() != null && composite.afterKey().isEmpty() == false;
     }
 
-    static CompositeAggregation getComposite(SearchResponse response) {
-        Aggregation agg = response.getAggregations().get(Aggs.ROOT_GROUP_NAME);
-        if (agg == null) {
-            return null;
-        }
+    static CompositeAggregationBuilder getCompositeBuilder(SearchSourceBuilder source) {
+        AggregationBuilder aggregation = source.aggregations()
+            .getAggregatorFactories()
+            .stream()
+            .filter(a -> Objects.equals(a.getName(), Aggs.ROOT_GROUP_NAME))
+            .findFirst()
+            .orElse(null);
 
-        if (agg instanceof CompositeAggregation) {
-            return (CompositeAggregation) agg;
-        }
+        Check.isTrue(aggregation instanceof CompositeAggregationBuilder, "Unexpected aggregation builder " + aggregation);
 
-        throw new SqlIllegalArgumentException("Unrecognized root group found; {}", agg.getClass());
+        return (CompositeAggregationBuilder) aggregation;
     }
 
-    private static void updateCompositeAfterKey(SearchResponse r, SearchSourceBuilder search) {
-        CompositeAggregation composite = getComposite(r);
+    static boolean mightProducePartialPages(CompositeAggregationBuilder aggregation) {
+        return aggregation.getPipelineAggregations().stream().anyMatch(a -> a instanceof BucketSelectorPipelineAggregationBuilder);
```

Comment:
I don't think we're using any other type of pipeline aggregation so simply returning if the size > 0 should be enough.
Also not a fun of using the streams api due to their cost - the above can become:
```
for (var a : aggregation.getPipelineAggregrations()) {
    return a instanceof BucketSelectorPipelineAggregationBuilder
}
```
```